renormalizer/cv/zerot.py

```diff
@@ -0,0 +1,351 @@
+                                           self.a_oper[isite - 1], first_R)
+            else:
+                path_a = [([0, 1], "abc, adef->bcdef"),
+                          ([3, 0], "bcdef, bdgh->cefgh"),
+                          ([2, 0], "cefgh, heij->cfgij"),
+                          ([1, 0], "cfgij, fjk->cgik")]
+                ax = multi_tensor_contract(path_a, first_L, Matrix(xstruct),
+                                           self.a_oper[isite - 2], self.a_oper[isite - 1],
+                                           first_R)
+            cout = ax[qnmat == constrain_qn].reshape(nonzeros, 1).asnumpy()
+            return cout
+
+        mat_a = scipy.sparse.linalg.LinearOperator((nonzeros, nonzeros), matvec=hop)
+        if num in [1, 2]:
+            x, info = scipy.sparse.linalg.cg(mat_a, vec_b.asnumpy(), atol=0)
+        else:
+            x, info = scipy.sparse.linalg.cg(mat_a, vec_b.asnumpy(), tol=1.e-5, x0=guess, M=pre_a_mat, atol=0)
+        logger.info('hop times:{count[0]}')
+        try:
+            assert info == 0
+        except:
+            logger.debug('iterative solver not converged')
+            pass
+        l_value = np.inner(hop(x).reshape(1, nonzeros), x.reshape(1, nonzeros)
+                     ) - 2 * np.inner(
+                         vec_b.reshape(1, nonzeros), x.reshape(1, nonzeros))
+        xstruct = solver.cvec2cmat(xshape, x, qnmat, constrain_qn)
+        x, xdim, xqn, compx = \
+            solver.renormalization_svd(xstruct, qnbigl, qnbigr, system,
+                                       constrain_qn, self.m_max, percent)
+        if self.method == "1site":
+            self.cv_mps[isite - 1] = x
+            if direction == "left":
+                if isite != 1:
+                    self.cv_mps[isite - 2] = tensordot(
+                        self.cv_mps[isite - 2], compx, axes=(-1, 0))
+                    self.cv_mps.qn[isite - 1] = xqn
+                else:
+                    self.cv_mps[isite - 1] = tensordot(
+                        compx, self.cv_mps[isite - 1], axes=(-1, 0))
+                    self.cv_mps.qn[isite - 1] = [0]
+            elif direction == "right":
+                if isite != len(self.cv_mps):
+                    self.cv_mps[isite] = tensordot(
+                        compx, self.cv_mps[isite], axes=(-1, 0))
+                    self.cv_mps.qn[isite] = xqn
+                else:
+                    self.cv_mps[isite - 1] = tensordot(
+                        self.cv_mps[isite - 1], compx, axes=(-1, 0))
+                    self.cv_mps.qn[isite] = [0]
+        else:
+            if direction == "left":
+                self.cv_mps[isite - 1] = x
+                self.cv_mps[isite - 2] = compx
+            else:
+                self.cv_mps[isite - 2] = x
+                self.cv_mps[isite - 1] = compx
+            self.cv_mps.qn[isite - 1] = xqn
+        return l_value
+
+    def initialize_LR(self, direction):
```

Comment:
I'm not sure about this, but have you looked at the `Environ` class in `renormalizer.mps.lib`? It is built to deal with this "LR" thing. If it does not satisfy your need, is it possible to make improvements to the `Environ` class so that other people can benefit from your code? Thanks.

Reply:
Good suggestions! It seems that the zero temperature structure looks much similar as ground state calculation, which will be easy by improving the Environ class a little bit.
However, for finite temperature code, firstly, it is not a <\psi|H\psi>-like structure, which could be more complicated to generalize a common structure for all terms. Any smart ideas?

![image](https://user-images.githubusercontent.com/36879153/64114223-06eaa780-cdbf-11e9-9a5b-4dcc3b608005.png)

Reply:
This is too complicated for me :stuck_out_tongue_closed_eyes:. I'd say let's leave it here, or we can hear what @jjren say about this. This PR requires his approvement anyway.

Reply:
If this kind of matrix contraction pattern shares some features with the other contraction pattern, I think it is better to put them together. However, if the pattern is independent and only for CV approach, it is ok to leave it here.

Reply:
Let's leave it here and we can refactor the code if we need a similar feature in the future.